### PR TITLE
fix(compatibility): fix AddressDetails in the customers flow

### PR DESCRIPTION
**Pull request by kritikmodi**
## Type of Change

- [x] Bugfix
- [x] Refactoring

## Description

Address of a new customer was being accepted as a `SecretSerdeValue` in the `CustomerCreate` and `CustomerUpdate` requests. This PR does the following : 

- Add a new struct `AddressDetails` in the `customers` flow for the `address` field.
- Add the `address` field in Stripe Compatibility Layer and the corresponding mappings for it.
- Add missing fields in `CustomerCreate` and `CustomerUpdate` requests as per Stripe API reference.

## Motivation and Context

This PR closes the following issues : 

## How did you test it?

Manually

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code

### CODE REVIEW

File: crates/api_models/src/customers.rs [hunk 1]
```diff
@@ -1,8 +1,52 @@
...
 use serde::{Deserialize, Serialize};
 use utoipa::ToSchema;
 
+use crate::enums as api_enums;
+
+impl SerializableSecret for AddressDetails {}
+
+#[derive(Debug, Default, Clone, Deserialize, Serialize, ToSchema)]
+#[serde(deny_unknown_fields)]
+pub struct AddressDetails {
```
Comment (NishantJoshi00) [inline crates/api_models/src/customers.rs:12]: what was the requirement of creating a new struct `AddressDetails`.

Is this any different from `crates/api_models/src/payments.rs:1134`

Comment (kritikmodi) [reply crates/api_models/src/customers.rs:12]: Nope, it's the same. But earlier 'customers` endpoint was taking `address` as `SecretSerdeValue` which was incorrect. So I just added this `AddressDetails` struct to take in only the specified values while making a request; the same way it's done for payments.

Comment (Abhicodes-crypto) [reply crates/api_models/src/customers.rs:12]: If it is the same as that in the payments.rs file, then don't create a duplicate struct, the same struct can be imported and used here.

File: crates/router/src/compatibility/stripe/customers/types.rs [hunk 2]
```diff
@@ -1,34 +1,86 @@
...
+}
 
 #[derive(Clone, Default, Serialize, Deserialize, PartialEq, Eq)]
 pub struct CreateCustomerRequest {
     pub email: Option<Email>,
     pub invoice_prefix: Option<String>,
-    pub name: Option<masking::Secret<String>>,
-    pub phone: Option<masking::Secret<String>>,
-    pub address: Option<masking::Secret<serde_json::Value>>,
+    pub name: Option<Secret<String>>,
```
Comment (Abhicodes-crypto) [inline crates/router/src/compatibility/stripe/customers/types.rs:42]: please don't use specific imports. If you see specific imports anywhere in compatibility layer feel free to change it.

**Review by Abhicodes-crypto — CHANGES_REQUESTED** [review]
Hope you have tested customer flow in both compatibility layer and core. please add screenshots also.

File: crates/router/src/compatibility/stripe/customers/types.rs [hunk 3]
```diff
@@ -1,34 +1,86 @@
...
+pub struct Shipping {
+    pub address: AddressDetails,
+    pub name: Option<Secret<String>>,
+    pub carrier: Option<String>,
+    pub phone: Option<Secret<String>>,
+    pub tracking_number: Option<Secret<String>>,
+}
+
+#[derive(Default, Serialize, PartialEq, Eq, Deserialize, Clone)]
+pub struct AddressDetails {
```
Comment (NishantJoshi00) [inline crates/router/src/compatibility/stripe/customers/types.rs:29]: can we have a different name for this? It seems to be conflicting with others and it's not enough to give context about why it was created

File: crates/router/src/compatibility/stripe/customers/types.rs [hunk 4]
```diff
@@ -1,34 +1,86 @@
...
 }
 
 #[derive(Clone, Default, Serialize, Deserialize, PartialEq, Eq)]
 pub struct CustomerUpdateRequest {
     pub description: Option<String>,
     pub email: Option<Email>,
-    pub phone: Option<masking::Secret<String, masking::WithType>>,
-    pub name: Option<masking::Secret<String>>,
-    pub address: Option<masking::Secret<serde_json::Value>>,
+    pub phone: Option<Secret<String, masking::WithType>>,
```
Comment (NishantJoshi00) [inline crates/router/src/compatibility/stripe/customers/types.rs:67]: if you are changing this, you don't have to mention `masking::WithType` it's auto-inferred.

File: crates/router/src/compatibility/stripe/customers/types.rs [hunk 5]
```diff
@@ -39,8 +91,8 @@ pub struct CreateCustomerResponse {
     pub description: Option<String>,
     pub email: Option<Email>,
     pub metadata: Option<pii::SecretSerdeValue>,
-    pub name: Option<masking::Secret<String>>,
-    pub phone: Option<masking::Secret<String, masking::WithType>>,
+    pub name: Option<Secret<String>>,
+    pub phone: Option<Secret<String, masking::WithType>>,
```
Comment (NishantJoshi00) [inline crates/router/src/compatibility/stripe/customers/types.rs:95]: same here

**Review by NishantJoshi00 — CHANGES_REQUESTED** [review]

File: crates/api_models/src/customers.rs (hunk 1 above)
Comment (kritikmodi) [reply crates/api_models/src/customers.rs:12]: Okay.

File: crates/router/src/compatibility/stripe/customers/types.rs (hunk 2 above)
Comment (kritikmodi) [reply crates/router/src/compatibility/stripe/customers/types.rs:42]: Okay

File: crates/router/src/compatibility/stripe/customers/types.rs [hunk 6]
```diff
@@ -134,9 +203,9 @@ pub struct PaymentMethodData {
 pub struct CardDetails {
     pub country: Option<String>,
     pub last4: Option<String>,
-    pub exp_month: Option<masking::Secret<String>>,
-    pub exp_year: Option<masking::Secret<String>>,
-    pub fingerprint: Option<masking::Secret<String>>,
+    pub exp_month: Option<Secret<String>>,
```
[the remark this answers is not in the document]
Comment (kritikmodi) [reply crates/router/src/compatibility/stripe/customers/types.rs:206]: Okay

File: crates/router/src/compatibility/stripe/customers/types.rs (hunk 4 above)
Comment (kritikmodi) [reply crates/router/src/compatibility/stripe/customers/types.rs:67]: Okay

File: crates/api_models/src/customers.rs [hunk 7]
```diff
@@ -70,18 +104,8 @@ pub struct CustomerResponse {
...
-    "line2": "Koramangala",
-    "line3": "Stallion",
-    "state": "Karnataka",
-    "zip": "560095",
-    "first_name": "John",
-    "last_name": "Doe"
-  }))]
-    pub address: Option<Secret<serde_json::Value>>,
+    #[schema(value_type = Option<AddressDetails>)]
+    pub address: Option<Secret<AddressDetails>>,
```
[the remark this answers is not in the document]
Comment (kritikmodi) [reply crates/api_models/src/customers.rs]: We're parsing it as a struct for documentation purposes mainly. Also, backward compatibility isn't needed here, so a struct would be more suited here I assume.

**Review by NishantJoshi00 — DISMISSED** [review]

**Review by inventvenkat — APPROVED** [review]

**Review by sahkal — APPROVED** [review]

**Review by inventvenkat — APPROVED** [review]

**Pull request merged**